File: P1/java/FactoriaCarretera.java
package java;
import java.util.*;

public class FactoriaCarretera implements FactoriaCarrerayBicicleta{

    private int numbicis;
    
    // Constructor
    public FactoriaCarretera(int n) {
        super();
        numbicis = n;
        if (n < 0)
            numbicis = 0;
    }

    // Metodo crear Carrera
    @Override
    public Carrera crearCarrera() {
        ArrayList<Bicicleta> bicis = new ArrayList<>();
        for(int i = 0; i < numbicis; i++){
            bicis.add(this.crearBicicleta());
        }
        return new CarreraCarretera(bicis);
    }

    // Metodo crear Bicicleta
    @Override
    public Bicicleta crearBicicleta() {
        return new BicicletaCarretera();
    }

}

File: P1/python/factoria_carretera.py
from factoria_carretera_bicicleta import FactoriaCarrerayBicicleta
from carrera_carretera import CarreraCarretera
from bicicleta_carrera import BicicletaCarretera
import copy

class FactoriaCarretera(FactoriaCarrerayBicicleta):
    def __init__(self,n):
        FactoriaCarrerayBicicleta.__init__(self)
        self.num_bicis=n 
        if n<0:
            self.num_bicis=0

    def crear_carrera(self):
        bicis = []
        for i in range(self.num_bicis):
            if(len(bicis)==0):
                bicis.append(self.crear_bicicleta())
            else:
                bicis.append(copy.deepcopy(bicis[0]))
            
        return CarreraCarretera(bicis) 

    def crear_bicicleta(self):
        return BicicletaCarretera()
